File: LR1/cpp/src/d/Stack.cpp
#include "d/Stack.hpp"
#include <fstream>

static const int MIN_CAPACITY = 4;

void initStack(Stack &s, int capacity) {
    if (capacity < MIN_CAPACITY) capacity = MIN_CAPACITY;
    s.data = new std::string[capacity];
    s.size = 0;
    s.capacity = capacity;
}

static void growIfNeeded(Stack &s) {
    if (s.size < s.capacity) return;
    int newCap = s.capacity * 2;
    std::string* newData = new std::string[newCap];
    for (int i = 0; i < s.size; ++i) newData[i] = s.data[i];
    delete[] s.data;
    s.data = newData;
    s.capacity = newCap;
}

static void shrinkIfNeeded(Stack &s) {
    if (s.capacity <= MIN_CAPACITY) return;
    if (s.size * 4 >= s.capacity) return;
    int newCap = s.capacity / 2;
    if (newCap < MIN_CAPACITY) newCap = MIN_CAPACITY;
    std::string* newData = new std::string[newCap];
    for (int i = 0; i < s.size; ++i) newData[i] = s.data[i];
    delete[] s.data;
    s.data = newData;
    s.capacity = newCap;
}

bool push(Stack &s, const std::string &val) {
    growIfNeeded(s);
    s.data[s.size++] = val;
    return true;
}

bool pop(Stack &s, std::string &val) {
    if (s.size == 0) return false;
    val = s.data[--s.size];
    shrinkIfNeeded(s);
    return true;
}

std::string top(Stack &s) {
    if (s.size == 0) return std::string();
    return s.data[s.size - 1];
}

int length(Stack &s) {
    return s.size;
}

void readFromFile(Stack &s, const std::string &filename) {
    std::ifstream in(filename);
    if (!in.is_open()) return;
    std::string val;
    while (in >> val) push(s, val);
    in.close();
}

void writeToFile(Stack &s, const std::string &filename) {
    std::ofstream out(filename);
    for (int i = 0; i < s.size; ++i) {
        out << s.data[i];
        if (i + 1 < s.size) out << " ";
    }
    out.close();
}

void clearStack(Stack &s) {
    delete[] s.data;
    s.data = nullptr;
    s.size = 0;
    s.capacity = 0;
}

File: LR1/rust/src/d/Stack.rs
use std::fs::File;
use std::io::{self, BufRead, Write};

const MIN_CAPACITY: usize = 4;

pub struct Stack {
    data: Vec<String>,
}

impl Stack {
    pub fn new() -> Self {
        Stack {
            data: Vec::with_capacity(MIN_CAPACITY),
        }
    }

    fn ensure_capacity(&mut self) {
        if self.data.len() == self.data.capacity() {
            self.data.reserve(self.data.capacity());
        }
    }

    fn shrink_if_needed(&mut self) {
        if self.data.capacity() > MIN_CAPACITY && self.data.len() * 4 < self.data.capacity() {
            let new_cap = self.data.capacity() / 2;
            self.data.shrink_to(new_cap.max(MIN_CAPACITY));
        }
    }

    pub fn push(&mut self, val: &str) {
        self.ensure_capacity();
        self.data.push(val.to_string());
    }

    pub fn pop(&mut self) -> Option<String> {
        let val = self.data.pop();
        self.shrink_if_needed();
        val
    }

    pub fn top(&self) -> Option<&String> {
        self.data.last()
    }

    pub fn len(&self) -> usize {
        self.data.len()
    }

    pub fn clear(&mut self) {
        self.data.clear();
    }

    pub fn read_from_file(&mut self, filename: &str) -> io::Result<()> {
        self.clear();
        let file = File::open(filename)?;
        for line in io::BufReader::new(file).lines() {
            for token in line?.split_whitespace() {
                self.push(token);
            }
        }
        Ok(())
    }

    pub fn write_to_file(&self, filename: &str) -> io::Result<()> {
        let mut file = File::create(filename)?;
        for (i, val) in self.data.iter().enumerate() {
            if i > 0 { write!(file, " ")?; }
            write!(file, "{}", val)?;
        }
        Ok(())
    }
}
